### ISSUE TITLE
MAX over empty set should return NULL according to standard SQL

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#13894]
**Describe the bug**
We break distributive properties for aggregations. For example  
```
max ( A U B ) ≠ max( max(A) U max(B) )
```

**How to reproduce**

``` sql
SELECT MAX(aggr)
FROM
(
    SELECT MAX(-1) AS aggr
    FROM system.one
    WHERE NOT 1
    UNION ALL
    SELECT MAX(-1) AS aggr
    FROM system.one
    WHERE 1
)

┌─MAX(aggr)─┐
│         0 │
└───────────┘
```

``` sql
SELECT MAX(-1)

┌─MAX(-1)─┐
│      -1 │
└─────────┘
```

``` sql

SELECT SUM(-1) AS aggr
FROM system.one
WHERE 0

┌─aggr─┐
│    0 │
└──────┘
```

**Expected behavior**
Aggregation should return NULL(but result is not nullable by default) or no rows.

**Additional context**
Found by SQLancer by ClickHouseTLPAggregateOracle

**Comment by den-crane** [comment]
Not a bug. Feature. As far as I remember there is an issue about a setting to enable automatic nullable aggregation results

```
SELECT SUM(cast(-1, 'Nullable(Int8)')) AS aggr
FROM system.one
WHERE 0

┌─aggr─┐
│ ᴺᵁᴸᴸ │
└──────┘
```

**Comment by qoega** [comment]
Nice. But actually can be very confusing when it encounters somewhere inside part of UNION ALL.
For ANSI compatibility mode we will have to enable it.

**Comment by qoega** [comment]
And `empty_result_for_aggregation_by_empty_set` is currently undocumented!

**Comment by alexey-milovidov** [comment]
@qoega `empty_result_for_aggregation_by_empty_set` does not make behaviour more standard compliant.
See №5 of https://github.com/ClickHouse/ClickHouse/issues/10540

**Comment by hczhcz** [comment]
It is `max( A U B )` vs `max( { max(A), max(B) } )`. This behavior is useful especially when the operator does not observe the law of association.

**Comment by alexey-milovidov** [comment]
Should be resolved with https://github.com/ClickHouse/ClickHouse/pull/16123.

**Comment by alexey-milovidov** [comment]
@ucasFL I thought it should start to work but it does not if MAX is written in capitals:

```
SELECT MAX(aggr)
FROM 
(
    SELECT MAX(-1) AS aggr
    FROM system.one
    WHERE NOT 1
    UNION ALL
    SELECT MAX(-1) AS aggr
    FROM system.one
    WHERE 1
)

Query id: b46f9344-cc6c-4545-8188-a74383826c1e


Received exception from server (version 20.12.1):
Code: 46. DB::Exception: Received from localhost:9000. DB::Exception: Unknown function MAXOrNull: While processing MAXOrNull(-1) AS aggr.
```

Although it works perfectly when canonical name is used.

**Comment by ucasFL** [comment]
@alexey-milovidov This is because MAX is case-insensative, however, maxOrNull is not, [#16774](https://github.com/ClickHouse/ClickHouse/pull/16774) may can resolve this problem.

```sql
:) select MAX(aggr) FROM (select MAX(-1) AS aggr FROM system.one where not 1 union all select MAX(-1) from system.one as aggr where 1 )

SELECT MAX(aggr)
FROM 
(
    SELECT MAX(-1) AS aggr
    FROM system.one
    WHERE NOT 1
    UNION ALL
    SELECT MAX(-1)
    FROM system.one AS aggr
    WHERE 1
)

Query id: 4086ea51-ce4d-4fbb-b640-ac65a76184fc

┌─maxOrNull(aggr)─┐
│              -1 │
└─────────────────┘

1 rows in set. Elapsed: 0.005 sec. 

```

**Comment by alexey-milovidov** [comment]
Fixed in version 20.12-testing when the setting `aggregate_functions_null_for_empty` is enabled.
